gtfs-validator/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/RouteUniqueNamesValidator.java

```diff
@@ -0,0 +1,87 @@
+package org.mobilitydata.gtfsvalidator.validator;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.annotation.Inject;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.notice.RouteUniqueNamesNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsRoute;
+import org.mobilitydata.gtfsvalidator.table.GtfsRouteTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsRouteType;
+
+/**
+ * Validates whether the combination of names, type and agency ID for a route is unique.
+ *
+ * <p>Generated notices: {@link RouteUniqueNamesNotice}.
+ */
+@GtfsValidator
+public class RouteUniqueNamesValidator extends FileValidator {
+  @Inject GtfsRouteTableContainer routeTable;
+
+  @Override
+  public void validate(NoticeContainer noticeContainer) {
+    final Map<Integer, List<Object>> testedRoutes = new HashMap<>();
+    for (GtfsRoute route : routeTable.getEntities()) {
+      // Grab necessary route information and transfer null to empty String to make later comparison
+      // and notice genration easier
+      final String routeLongName = Optional.ofNullable(route.routeLongName()).orElse("");
+      final String routeShortName = Optional.ofNullable(route.routeShortName()).orElse("");
+      final GtfsRouteType routeType = route.routeType();
+      final String agencyId = Optional.ofNullable(route.agencyId()).orElse("");
+
+      final int hashCode = Objects.hash(routeLongName, routeShortName, routeType, agencyId);
+      if (testedRoutes.containsKey(hashCode)) {
+        List<Object> storedRouteInfo = testedRoutes.get(hashCode);
+        // Ensure routes with the same hashcode do have the same route names, types and agency IDs
+        if (routeLongName.equals(storedRouteInfo.get(1))
+            && routeShortName.equals(storedRouteInfo.get(2))
+            && routeType.getNumber() == (int) (storedRouteInfo.get(3))
+            && agencyId.equals(storedRouteInfo.get(4))) {
+          noticeContainer.addNotice(
+              new RouteUniqueNamesNotice(
+                  route.routeId(),
+                  route.csvRowNumber(),
+                  /* comparedRouteId= */ storedRouteInfo.get(0).toString(),
+                  /* comparedRouteCsvRowNumber= */ (long) (storedRouteInfo.get(5)),
+                  routeLongName,
+                  routeShortName,
+                  routeType,
+                  agencyId));
+        }
+      } else {
+        List<Object> routeInfo =
+            new ArrayList<>(
+                Arrays.asList(
+                    route.routeId(),
+                    routeLongName,
+                    routeShortName,
+                    routeType.getNumber(),
+                    agencyId,
+                    route.csvRowNumber()));
```

Comment:
The code could be much more simplified if you used something like:

@AutoValue
abstract class RouteIdentifier {
    static RouteIdentifier create(String shortName,String longName, GtfsRouteType routeType, String agencyId) {
        return new AutoValue_RouteIdentifier(shortName, longName,routeType,agencyId);
    }

    abstract String shortName();
    abstract String longName();
    abstract GtfsRouteType routeType();
    abstract String agencyId();
}

Reply:
Thank you for the advice! I am considering that.

Reply:
Done. I have used autovalue now.